shared/src/metabase/shared/util/time.cljc

```diff
@@ -0,0 +1,255 @@
+         ;; This mimics Moment.js behavior if you `m.startOf("unknown unit")` - it doesn't change anything.
+         [value value])
+       (defmethod to-range "minute" [value _]
+         (let [start (t/truncate-to value :minutes)]
+           [start (minus-ms (t/plus start (t/minutes 1)))]))
+       (defmethod to-range "hour" [value _]
+         (let [start (t/truncate-to value :hours)]
+           [start (minus-ms (t/plus start (t/hours 1)))]))
+       (defmethod to-range "day" [value _]
+         (let [start (t/truncate-to value :days)]
+           [start (minus-ms (t/plus start (t/days 1)))]))
+       (defmethod to-range "week" [value _]
+         (let [first-day (first-day-of-week)
+               start (-> value
+                         (t/truncate-to :days)
+                         (t/adjust :previous-or-same-day-of-week first-day))]
+           [start (minus-ms (t/plus start (t/weeks 1)))]))
+       (defmethod to-range "month" [value _]
+         (let [value (t/truncate-to value :days)]
+           [(t/adjust value :first-day-of-month)
+            (minus-ms (t/adjust value :first-day-of-next-month))]))
+       (defmethod to-range "year" [value _]
+         (let [value (t/truncate-to value :days)]
+           [(t/adjust value :first-day-of-year)
+            (minus-ms (t/adjust value :first-day-of-next-year))]))))
+
+(defmulti ^:private parse-timestamp-text by-unit)
+
+(defmethod parse-timestamp-text :default [value _]
+  ;; Best effort to parse this unknown string format, as a local zoneless datetime, then treating it as UTC.
+  #?(:cljs (moment/utc value)
+     :clj  (let [base (try (t/local-date-time value)
+                           (catch Exception _
+                             (try (t/local-date value)
+                                  (catch Exception _
+                                    nil))))]
+             (when base
+               (t/offset-date-time base (t/zone-id))))))
+
+(defmethod parse-timestamp-text "day-of-week" [value options]
+  ;; Try to parse as a regular timestamp; if that fails then try to treat it as a weekday name and adjust from
+  ;; the current time.
+  (let [as-default (try ((get-method parse-timestamp-text :default) value options)
+                        (catch #?(:clj Exception :cljs js/Error) _
+                          nil))]
+    (if (valid-datetime? as-default)
+      as-default
+      #?(:cljs (-> (now)
+                   (.isoWeekday value)
+                   (.startOf "day"))
+         :clj  (let [day (try (t/day-of-week "EEE" value)
+                              (catch Exception _
+                                (t/day-of-week "EEEE" value)))]
+                 (-> (now)
+                     (t/truncate-to :days)
+                     (t/adjust :previous-or-same-day-of-week :monday)  ; Move to ISO start of week.
+                     (t/adjust :next-or-same-day-of-week day)))))))    ; Then to the specified day.
+
+(defmulti ^:private parse-timestamp-number by-unit)
```

Comment:
I actually like that a lot better. The JS still calls it "parse", but that's okay. I've updated all of this to `coerce`.

There's docs now, since it's a public var `common/coerce-timestamp-number`. I've expanded the `coerce-timestamp-number` ones to explain when the magic date is used (only for relative numbers, `day-of-year` and `day-of-month`).